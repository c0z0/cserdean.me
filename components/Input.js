import React from "react";
import styled, { keyframes } from "styled-components";
import PropTypes from "prop-types";

const InputWrapper = styled.div`
  position: relative;
  display: inline-block;
  margin-top: 16px;
  margin-bottom: 32px;
`;

const InputClear = styled.button.attrs({
  type: "button",
  "data-testid": "input-clear-button"
})`
  background: none;
  border: none;
  outline: none;
  color: white;
  cursor: pointer;
  position: absolute;
  font-size: 1.5em;
  right: -5px;
  top: 50%;
  transform: translate(0, -50%) rotate(45deg);
`;

const LoadingKeyframes = keyframes`
 0% {
      left: 0;
      right: 75%;
    }
    50% {
      left: 75%;
      right: 0;
    }
    100% {
      left: 0;
      right: 75%;
    }
`;

const Loading = styled.span`
  position: absolute;
  background: white;
  bottom: 0px;
  height: 1px;
  animation: ${LoadingKeyframes} 2s infinite;
`;

const InputElement = styled.input.attrs({
  "data-testid": "input-element",
  type: "text",
  placeholder: "Type something..."
})`
  width: 290px;
  padding-right: 10px;
  height: 18px;
  border: none;
  border-bottom: solid rgba(255, 255, 255, 0.4) 1px;
  padding: 8px 0;
  outline: none;
  font-size: 16px;
  background: none;
  transition: 0.2s all;
  color: white;
  padding-right: 10px;

  &:focus {
    border-color: rgba(
      255,
      255,
      255,
      ${({ loading }) => (loading ? "0.4" : "1")}
    );
  }
`;

export default function Input({ loading, text, onClear, onChange }) {
  return (
    <InputWrapper>
      {text.length > 0 && <InputClear onClick={onClear}>+</InputClear>}
      <InputElement loading={loading} value={text} onChange={onChange} />
      {loading && <Loading data-testid="input-loading" />}
    </InputWrapper>
  );
}

Input.propTypes = {
  loading: PropTypes.bool.isRequired,
  text: PropTypes.string.isRequired,
  onClear: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired
};
