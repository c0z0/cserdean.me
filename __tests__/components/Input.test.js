import React from "react";
import { render, cleanup, fireEvent } from "react-testing-library";
import Input from "../../components/Input";
import "jest-styled-components";

afterEach(cleanup);

test("Matches snapshot", () => {
  const { container, rerender } = render(
    <Input text="" loading onChange={() => {}} onClear={() => {}} />
  );

  expect(container.firstChild).toMatchSnapshot();

  rerender(
    <Input text="a" loading={false} onChange={() => {}} onClear={() => {}} />
  );

  expect(container.firstChild).toMatchSnapshot();
});

test("Clear button triggers", () => {
  const clear = jest.fn();

  const { getByTestId } = render(
    <Input text="a" loading onChange={() => {}} onClear={clear} />
  );

  fireEvent.click(getByTestId("input-clear-button"));

  expect(clear).toBeCalled();
});

test("On change event triggers", () => {
  const change = jest.fn();

  const { getByTestId } = render(
    <Input text="a" loading onChange={change} onClear={() => {}} />
  );

  fireEvent.change(getByTestId("input-element"), { target: { value: "b" } });

  expect(change).toBeCalled();
});
