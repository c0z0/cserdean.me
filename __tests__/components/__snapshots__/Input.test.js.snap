// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
.c0 {
  position: relative;
  display: inline-block;
  margin-top: 16px;
  margin-bottom: 32px;
}

.c2 {
  position: absolute;
  background: white;
  bottom: 0px;
  height: 1px;
  -webkit-animation: iHpxMo 2s infinite;
  animation: iHpxMo 2s infinite;
}

.c1 {
  width: 290px;
  height: 18px;
  border: none;
  border-bottom: solid rgba(255,255,255,0.4) 1px;
  padding: 8px 0;
  padding-right: 10px;
  outline: none;
  font-size: 16px;
  background: none;
  -webkit-transition: 0.2s all;
  transition: 0.2s all;
  color: white;
}

.c1:focus {
  border-color: rgba( 255, 255, 255, 0.4 );
}

<div
  class="c0"
>
  <input
    class="c1"
    data-testid="input-element"
    placeholder="Type something..."
    type="text"
    value=""
  />
  <span
    class="c2"
    data-testid="input-loading"
  />
</div>
`;

exports[`Matches snapshot 2`] = `
<div
  class="sc-bdVaJa drqCWm"
>
  .c0 {
  background: none;
  border: none;
  outline: none;
  color: white;
  cursor: pointer;
  position: absolute;
  font-size: 1.5em;
  right: -5px;
  top: 50%;
  -webkit-transform: translate(0,-50%) rotate(45deg);
  -ms-transform: translate(0,-50%) rotate(45deg);
  transform: translate(0,-50%) rotate(45deg);
}

<button
    class="c0"
    data-testid="input-clear-button"
    type="button"
  >
    +
  </button>
  <input
    class="sc-bxivhb hezTQd"
    data-testid="input-element"
    placeholder="Type something..."
    type="text"
    value="a"
  />
</div>
`;
